### PR TITLE
[#25] Add lawful Semigroup and Monoid instances

### LINKED ISSUE #25: Add lawful `Semigroup` and `Monoid` instances to `Treap` and `RTreap`
I think `merge` can be an implementation of `<>`.

### PR DESCRIPTION
**Pull request by sphaso**
Resolves #25

### CODE REVIEW

File: src/Treap/Pure.hs [hunk 1]
```diff
@@ -82,6 +82,12 @@ data Treap m a
     deriving stock (Show, Read, Eq, Generic, Foldable)
     deriving anyclass (NFData)
 
+instance Measured m a => Semigroup (Treap m a) where
```
Comment (chshersh) [inline src/Treap/Pure.hs:87]: Let's add documentation to this instance saying that `<>` is implemented via `merge`.

File: test/Test/Rand/Laws.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+monoidSpec :: Spec
+monoidSpec =
+    describe "Monoid" $
+        it "identity" $
+            require $ property $ do
+                a :: TestTreap <- fromList <$> forAll prop_int_list
+                let
+                    e = emptyTreap
+                a <> e === a
```
Comment (chshersh) [inline test/Test/Rand/Laws.hs]: I think you don't need `e` assigned to `emptyTreap` at all. You can have just two tests here:

```haskell
a <> mempty === a
mempty <> a === a
```

File: treap.cabal [hunk 3]
```diff
@@ -79,10 +79,13 @@ test-suite treap-test
                          Test.Rand.Cuts
                          Test.Rand.Query
                          Test.Rand.Update
+                         Test.Rand.Laws
 
   build-depends:       treap
                      , hspec
                      , hspec-core
+                     , hedgehog >= 0.6
+                     , hw-hspec-hedgehog
```
Comment (chshersh) [inline treap.cabal:88]: Nice, I haven't heard about this library. Integrating `hedgehog` with `hspec` can make things simpler

File: treap.cabal [hunk 4]
```diff
@@ -79,10 +79,13 @@ test-suite treap-test
                          Test.Rand.Cuts
                          Test.Rand.Query
                          Test.Rand.Update
+                         Test.Rand.Laws
 
   build-depends:       treap
                      , hspec
                      , hspec-core
+                     , hedgehog >= 0.6
```
Comment (chshersh) [inline treap.cabal]: It's always a good idea to add upper and lower bounds
```suggestion
                     , hedgehog ^>= 0.6
```

**Review by chshersh — COMMENTED** [review]
Thanks for the change, it looks good! I have some comments but mostly nice

File: treap.cabal (hunk 3 above)
Comment (sphaso) [reply treap.cabal:88]: it's a pretty new library but I figured it was better than adding tasty just to run hedgehog :)

**Comment by sphaso** [comment]
@chshersh fixed the test and added docs :+1:

**Review by chshersh — APPROVED** [review]
Thanks for your work!

**Pull request merged**